rpc/rpc.go

```diff
@@ -1517,24 +1514,26 @@ func (c *lockedConn) recvPayload(rl *releaseList, payload rpccp.Payload) (_ capn
 		c.er.ReportError(exc.WrapError("read payload: capability table", err))
 		return p, nil, nil
 	}
-	mtab := make([]capnp.Client, ptab.Len())
+
+	mtab := payload.Message().CapTable()
+	mtab.Reset()
+
+	var cl capnp.Client
 	for i := 0; i < ptab.Len(); i++ {
-		var err error
-		mtab[i], err = c.recvCap(ptab.At(i))
-		if err != nil {
-			for _, client := range mtab[:i] {
-				rl.Add(client.Release)
-			}
-			return capnp.Ptr{}, nil, rpcerr.Annotate(
-				err, "read payload: capability "+str.Itod(i),
-			)
+		if cl, err = c.recvCap(ptab.At(i)); err != nil {
+			cl.Release()
```

Comment:
It is not safe to release clients while holding the connection lock; it could deadlock. This is why we passed in a releaseList.